app/models/user.rb

```diff
@@ -200,123 +229,72 @@ def can_make_challenges
   end
 
+  ### Moderation ###
   def approve(reason, moderator)
-    ModeratorLog.create(user_id: moderator.id,
-                        target: self,
-                        action: "#{moderator.username} has approved #{username}.",
-                        reason: reason)
-    update_attribute(:approved, true)
+    create_mod_log(moderator, "#{moderator.username} has approved #{username}.", reason)
+    update_retain_password(approved: true)
+  end
+
+  def latest_ban
+    site_ban.find_by("'#{Time.now.utc}' < expiration")
   end
 
   def lift_ban(reason, moderator)
     # Do not lift bans on users marked for death.
-    unless marked_for_death
-      site_ban = SiteBan.find_by("'#{Time.now.utc}' < expiration AND user_id = #{id}")
-      unless site_ban.nil?
-        ModeratorLog.create(user_id: moderator.id,
-                            target: self,
-                            action: "#{moderator.username} has lifted #{username}'s ban (original expiry: #{ApplicationController.helpers.date_string_short(site_ban.expiration)}).",
-                            reason: reason)
-        site_ban.destroy
-      end
-    end
+    return if marked_for_death || latest_ban.nil?
+
+    action = "#{moderator.username} has lifted #{username}'s ban " \
+              "(original expiry: #{latest_ban.expiration})."
+    create_mod_log(moderator, action, reason)
+    latest_ban.update(expiration: Time.now.utc.to_date)
   end
 
   def ban_user(duration, reason, moderator)
     # Bans cannot be modified after a user has been marked for death.
-    unless marked_for_death
-      # Find and update an existing ban.
-      site_ban = SiteBan.find_by("'#{Time.now.utc}' < expiration AND user_id = #{id}")
-      if site_ban.nil?
-        # Or create a new one if not found.
-        SiteBan.create(user_id: id,
-                       expiration: Time.now.utc.to_date + duration.to_i.days,
-                       reason: reason)
-        ModeratorLog.create(user_id: moderator.id,
-                            target: self,
-                            action: "#{moderator.username} has banned #{username} for #{duration} days.",
-                            reason: reason)
-      else
-        site_ban.expiration = Time.now.utc.to_date + duration.days
-        site_ban.reason = reason
-        site_ban.save
-        ModeratorLog.create(user_id: moderator.id,
-                            target: self,
-                            action: "#{moderator.username} has adjusted #{username}'s ban to #{duration} days.",
-                            reason: reason)
-      end
-    end
+    return if marked_for_death
+
+    sb = latest_ban || site_ban.new
+    action = sb.new_record? ? " was banned for #{duration}" : "'s ban was adjusted to #{duration}"
+    create_mod_log(moderator, username + action + "days by #{moderator.username}", reason)
+    sb.update(expiration: duration.to_i.days.from_now.to_date, reason: reason)
```

Comment:
Does .update save properly if the ActiveRecord instance being referred to is a new one?

Reply:
Yes